internal/ethereum/confirmation_reconciler.go

```diff
@@ -52,71 +59,238 @@ func (bl *blockListener) reconcileConfirmationsForTransaction(ctx context.Contex
-	newConfirmationsWithoutTxBlock, lastValidatedBlock, err := bl.buildConfirmationQueueUsingInMemoryPartialChain(ctx, txBlockInfo, targetConfirmationCount)
+func (s *splice) hasGap() bool {
+	return len(s.earlyList) > 0 &&
+		len(s.lateList) > 0 &&
+		s.earlyList[len(s.earlyList)-1].BlockNumber.Uint64()+1 < s.lateList[0].BlockNumber.Uint64()
+}
+
+func (s *splice) containsTransactionBlock() bool {
+	// we haven't removed the first block from the early list
+	return len(s.earlyList) > 0
+}
+
+func (s *splice) fillGap(ctx context.Context, blockListener *blockListener) error {
+	if !s.hasGap() {
+		// no gap to fill
+		return nil
+	}
+
+	// fill one slot in the gap between the late list and the early list
+	// always fill from the end of the gap ( i.e. the block before the start of the late list) because
+	// the late list is our best view of the current canonical chain so working backwards from there will increase the number of blocks that we have a high confidence in
+
+	freshBlockInfo, _, err := blockListener.getBlockInfoByNumber(ctx, s.lateList[0].BlockNumber.Uint64()-1, false, "", "")
 	if err != nil {
-		return nil, err
+		return err
+	}
+	if freshBlockInfo == nil {
+		return i18n.NewError(ctx, msgs.MsgBlockNotAvailable)
 	}
 
-	// Special case: if targetConfirmationCount is 0, transaction is immediately confirmed
-	if targetConfirmationCount == 0 {
-		reconcileResult.Confirmations = []*ffcapi.MinimalBlockInfo{txBlockInfo}
-		reconcileResult.Confirmed = true
-		return reconcileResult, nil
+	fetchedBlock := &ffcapi.MinimalBlockInfo{
+		BlockNumber: fftypes.FFuint64(freshBlockInfo.Number.BigInt().Uint64()),
+		BlockHash:   freshBlockInfo.Hash.String(),
+		ParentHash:  freshBlockInfo.ParentHash.String(),
 	}
 
-	// Validate existing confirmations and fill gaps in the confirmation queue
-	var confirmations []*ffcapi.MinimalBlockInfo
-	var newFork bool
-	confirmations, newFork, err = bl.checkAndFillInGap(ctx, newConfirmationsWithoutTxBlock, existingConfirmations, txBlockInfo, targetConfirmationCount, lastValidatedBlock)
+	// Validate parent-child relationship
+	if !fetchedBlock.IsParentOf(s.lateList[0]) {
+		// most likely explanation of this is an unstable chain
+		return i18n.NewError(ctx, msgs.MsgFailedToBuildConfirmationQueue)
+	}
+
+	// Prepend fetched block to confirmation queue
+	s.lateList = append([]*ffcapi.MinimalBlockInfo{fetchedBlock}, s.lateList...)
+	return nil
+}
+
+func (s *splice) removeBrokenLink() {
+	if s.hasGap() {
+		// nothing to remove if there is a gap
+		return
+	}
+	// remove the last block from the early list because it is not the parent of the first block in the late list and we have higher confidence in the late list
+	s.earlyList = s.earlyList[:len(s.earlyList)-1]
+
+}
+
+func (s *splice) toSingleLinkedList() []*ffcapi.MinimalBlockInfo {
+	if s.hasGap() {
+		return nil
+	}
+	if len(s.earlyList) == 0 {
+		return s.lateList
+	}
+	if len(s.lateList) == 0 {
+		return s.earlyList
```

Comment:
This line feels dangerous as `lateList` seems to be the `validated` (against the latest chain state) list, if it's empty, then the last element of `earlyList` should be at least validated against the chain.

I think the logic is OK given `createLateList` will guarantee there will be at least 1 element in the list, and the for loop doesn't ever remove element from `lateList`